src/Features/Core/Portable/CodeFixesAndRefactorings/FixAllCodeAction.cs

```diff
@@ -0,0 +1,92 @@
+    {
+        internal readonly IFixAllState FixAllState;
+        private bool _showPreviewChangesDialog;
+
+        protected FixAllCodeAction(
+            IFixAllState fixAllState, bool showPreviewChangesDialog)
+        {
+            FixAllState = fixAllState;
+            _showPreviewChangesDialog = showPreviewChangesDialog;
+        }
+
+        protected abstract bool IsInternalProvider(IFixAllState fixAllState);
+        protected abstract IFixAllContext CreateFixAllContext(IFixAllState fixAllState, IProgressTracker progressTracker, CancellationToken cancellationToken);
+
+        public override string Title
+            => this.FixAllState.Scope switch
+            {
+                FixAllScope.Document => FeaturesResources.Document,
+                FixAllScope.Project => FeaturesResources.Project,
+                FixAllScope.Solution => FeaturesResources.Solution,
+                FixAllScope.ContainingMember => FeaturesResources.Containing_Member,
+                FixAllScope.ContainingType => FeaturesResources.Containing_Type,
+                _ => throw ExceptionUtilities.UnexpectedValue(this.FixAllState.Scope),
+            };
+
+        internal override string Message => FeaturesResources.Computing_fix_all_occurrences_code_fix;
+
+        protected override async Task<IEnumerable<CodeActionOperation>> ComputeOperationsAsync(CancellationToken cancellationToken)
+            => await ComputeOperationsAsync(new ProgressTracker(), cancellationToken).ConfigureAwait(false);
+
+        internal override Task<ImmutableArray<CodeActionOperation>> ComputeOperationsAsync(
+            IProgressTracker progressTracker, CancellationToken cancellationToken)
+        {
+            cancellationToken.ThrowIfCancellationRequested();
+            FixAllLogger.LogState(FixAllState, IsInternalProvider(FixAllState));
+
+            var service = FixAllState.Project.Solution.Workspace.Services.GetRequiredService<IFixAllGetFixesService>();
+
+            var fixAllContext = CreateFixAllContext(FixAllState, progressTracker, cancellationToken);
+            progressTracker.Description = fixAllContext.GetDefaultFixAllTitle();
+
+            return service.GetFixAllOperationsAsync(fixAllContext, _showPreviewChangesDialog);
+        }
+
+        internal sealed override Task<Solution?> GetChangedSolutionAsync(
+            IProgressTracker progressTracker, CancellationToken cancellationToken)
+        {
+            cancellationToken.ThrowIfCancellationRequested();
+            FixAllLogger.LogState(FixAllState, IsInternalProvider(FixAllState));
+
+            var service = FixAllState.Project.Solution.Workspace.Services.GetRequiredService<IFixAllGetFixesService>();
+
+            var fixAllContext = CreateFixAllContext(FixAllState, progressTracker, cancellationToken);
+            progressTracker.Description = fixAllContext.GetDefaultFixAllTitle();
+
+            return service.GetFixAllChangedSolutionAsync(fixAllContext);
+        }
```

Comment:
something is weird here having ComputeOperationsAsync and GetChangedSolutionAsync right?  One feels very redundant witht eh other (esp. as their impls are practically identicaly).  For these common FixAlls, should ComputeOperationsAsync just wrap GetChangedSolutionAsync and then wrap that in an ApplyChangesOperation?

Reply:
I agree, but I am going to avoid any logic changes in this existing code to avoid unnecessary regressions here. We can do this cleanup with a follow-up PR.